src/compas_fab/planning/action.py

```diff
@@ -0,0 +1,765 @@
+                Transformation.from_frame(robot_state.frame) * attached_tool_state.attached_to_robot_grasp
+            )
+            if debug:
+                print("- Attached Tool %s Followed." % attached_tool_id)
+
+        # Transform attached workpieces
+        attached_workpiece_id = scene_state.get_attached_workpiece_id()
+        if attached_workpiece_id is not None:
+            attached_workpiece = scene_state.get_workpiece_state(attached_workpiece_id)
+            attached_workpiece.frame = Frame.from_transformation(
+                Transformation.from_frame(robot_state.frame) * attached_workpiece.attached_to_robot_grasp
+            )
+            if debug:
+                print("- Attached Workpiece %s Followed." % attached_workpiece_id)
+
+
+class LinearMotion(RoboticAction):
+    """Action class to describe a Linear robotic movement.
+
+    A linear robotic movement moves the robot flange linearly in Cartesian space. The motion can
+    contain multiple linear segments (see intermediate_targets attribute).
+
+    Typically, the orientation between the starting frame and the target frame is the same.
+    However some linear motion planners can also interpolate between different orientations, check
+    the documentation of the planner for more details.
+
+
+    Attributes
+    ----------
+    intermediate_targets : list(:class:`compas.geometry.Frame`), optional
+        List of frames to define a linear movement that has multiple linear segments.
+        The frames are specified in the world coordinate frame.
+        Note that only the intermediate targets are specified, the starting and ending frames
+        should not be included in this list. The ending frame is specified by the target_robot_flange_frame attribute.
+
+        Default is an empty list, meaning the movement only has a single linear segment.
+
+    see :class:`compas_fab.planning.RoboticAction` for other attributes.
+    """
+
+    def __init__(self):
+        super(LinearMotion, self).__init__()
+        self.intermediate_targets = []  # type: Optional[list[Point]]
+        self.tag = "Linear Movement"
+
+    @property
+    def data(self):
+        data = super(LinearMotion, self).data
+        data["intermediate_targets"] = self.intermediate_targets
+        return data
+
+    @data.setter
+    def data(self, data):
+        super(LinearMotion, type(self)).data.fset(self, data)
+        self.intermediate_targets = data.get("intermediate_targets", self.intermediate_targets)
+
+
+class FreeMotion(RoboticAction):
+    """Action class for free robotic movements.
+    Free robotic movements are planned by Free Motion Planners.
+
+    Attributes
+    ----------
+    intermediate_planning_waypoint : list(:class:`compas.robots.Configuration`)
+        List of configurations to define a multi-step free movement.
+        The first and last configuration (the starting and target configuration) is not required.
+        Only include the intermediate waypoints.
```

Comment:
The need for this comment is exactly what makes it odd, it's not really intuitive that you need to strip parts of the data for this input param.